core-processor/src/processing.rs

```diff
@@ -147,11 +147,49 @@ where
                 res.system_reservation_context,
                 ActorExecutionErrorReplyReason::Trap(reason),
             ),
-            DispatchResultKind::Success => process_success(Success, res),
+            DispatchResultKind::Success => {
+                if let (Some(current_prev), Some(current_res)) = (
+                    system_reservation_ctx.current_reservation,
+                    res.system_reservation_context.current_reservation,
+                ) {
+                    debug_assert!(current_res <= current_prev);
```

Comment:
I'd add an explanotary comment, as it seems you are asserting the invariant, aren't you?
also, maybe even have a separate match arm for variants, for which reservations must be checked. That could reduce copy-paste I guess